File: authentication/typescript/azure/src/azureIdentityAuthenticationProvider.ts
import { GetTokenOptions, TokenCredential } from "@azure/core-auth";
import { BaseBearerTokenAuthenticationProvider } from '@microsoft/kiota-abstractions';
import { AzureIdentityAccessTokenProvider } from "./azureIdentityAccessTokenProvider";

export class AzureIdentityAuthenticationProvider extends BaseBearerTokenAuthenticationProvider {
    /**
     *@constructor
    *@param credentials The tokenCredential implementation to use for authentication.
    *@param scopes The scopes to use for authentication.
    *@param options The options to use for authentication.
    *@param allowedHosts The allowed hosts to use for authentication.
    */
    public constructor(credentials: TokenCredential,
        scopes: string[] = ['https://graph.microsoft.com/.default'],
        options?: GetTokenOptions,
        allowedHosts: Set<string> = new Set<string>(['graph.microsoft.com', 'graph.microsoft.us', 'dod-graph.microsoft.us', 'graph.microsoft.de', 'microsoftgraph.chinacloudapi.cn', 'canary.graph.microsoft.com'])) {
        super(new AzureIdentityAccessTokenProvider(credentials, scopes, options, allowedHosts));
    }
  }
  public getAuthorizationToken = async (
    _: RequestInformation
  ): Promise<string> => {
    const result = await this.credentials.getToken(this.scopes);
    return result?.token ?? "";
  };
}

File: authentication/go/azure/azure_identity_authentication_provider.go
// Package microsoft_kiota_authentication_azure implements Kiota abstractions for authentication using the Azure Core library.
// In order to use this package, you must also add the github.com/Azure/azure-sdk-for-go/sdk/azidentity.
package microsoft_kiota_authentication_azure

import (
	"context"
	"errors"

	azcore "github.com/Azure/azure-sdk-for-go/sdk/azcore"
	azpolicy "github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
	abs "github.com/microsoft/kiota/abstractions/go"
	auth "github.com/microsoft/kiota/abstractions/go/authentication"
)

// AzureIdentityAuthenticationProvider implementation of AuthenticationProvider that supports implementations of TokenCredential from Azure.Identity.
type AzureIdentityAuthenticationProvider struct {
	auth.BaseBearerTokenAuthenticationProvider
	scopes     []string
	credential azcore.TokenCredential
}

// NewAzureIdentityAuthenticationProvider creates a new instance of the AzureIdentityAuthenticationProvider using "https://graph.microsoft.com/.default" as the default scope.
func NewAzureIdentityAuthenticationProvider(credential azcore.TokenCredential) (*AzureIdentityAuthenticationProvider, error) {
	return NewAzureIdentityAuthenticationProviderWithScopes(credential, nil)
}

// NewAzureIdentityAuthenticationProviderWithScopes creates a new instance of the AzureIdentityAuthenticationProvider.
func NewAzureIdentityAuthenticationProviderWithScopes(credential azcore.TokenCredential, scopes []string) (*AzureIdentityAuthenticationProvider, error) {
	if credential == nil {
		return nil, errors.New("credential cannot be nil")
	}
	scopesLen := len(scopes)
	finalScopes := make([]string, scopesLen)
	if scopesLen == 0 {
		finalScopes = append(finalScopes, "https://graph.microsoft.com/.default")
	} else {
		copy(finalScopes, scopes)
	}
	baseBearer := auth.NewBaseBearerTokenAuthenticationProvider(
		func(request abs.RequestInformation) (string, error) {
			options := azpolicy.TokenRequestOptions{
				Scopes: finalScopes,
			}
			token, err := credential.GetToken(context.Background(), options)
			if err != nil {
				return "", err
			}
			return token.Token, nil
		})
	result := &AzureIdentityAuthenticationProvider{
		BaseBearerTokenAuthenticationProvider: *baseBearer,
		credential:                            credential,
		scopes:                                finalScopes,
	}

	return result, nil
}
